### PR TITLE
Implement River, open melds, and update Yaku logic

### PR DESCRIPTION
**Pull request by applyuser160**
This PR introduces support for tracking discards via the `River` struct and managing open melds (Chii, Pon, Kan) in a player's `Hand`. It also updates the `yaku` evaluation logic to differentiate between closed and open hands correctly. Crucially, it handles the complex edge cases related to winning via Ron for counting closed triplets in Yakus like `Sanankou` and `Suuankou`.

All additions follow TDD principles and include comprehensive test coverage. `cargo test` and `cargo clippy` report no warnings or errors.

---
*PR created automatically by Jules for task [4336679152619416454](https://jules.google.com/task/4336679152619416454) started by @applyuser160*

### CODE REVIEW

File: src/mahjong/hand.rs [hunk 1]
```diff
@@ -35,4 +44,56 @@ impl Hand {
         self.len -= 1;
         removed
     }
+
+    pub fn call_meld(
```
Comment (applyuser160) [inline src/mahjong/hand.rs]: @jules
この仕様だと順子の場合、鳴いた対象の牌は保持できるが、それによって作られた面子を構成するほかの牌がどれなのかが保持できない気がします

Comment (applyuser160) [reply src/mahjong/hand.rs]: @jules
修正内容は問題ないと思いますが、その前に、こちらでブランチに変更を加えたコミットを入れているので、そのコミットを踏まえた変更にしてください
